src/wfs/src/main/java/org/geoserver/wfs/request/FeatureCollectionResponse.java

```diff
@@ -53,15 +62,84 @@ public void setTimeStamp(Calendar timeStamp) {
         eSet(adaptee, "timeStamp", timeStamp);
     }
 
+    /**
+     * Factory method creating a new FeatureCollectionResponse.
+     *
+     * @return feature collection response
+     */
     public abstract FeatureCollectionResponse create();
 
+    /**
+     * Number of features included in this response. Number reflect the number of features included,
+     * which may be less than {@link #getTotalNumberOfFeatures()} when paging through more content
+     * than can be obtained in a single request.
+     *
+     * @return number of features included in this response.
+     */
     public abstract BigInteger getNumberOfFeatures();
 
+    /**
+     * Number of features included in response.
+     *
+     * @param n Number of features included in response
+     */
     public abstract void setNumberOfFeatures(BigInteger n);
 
-    public abstract BigInteger getTotalNumberOfFeatures();
+    /**
+     * Used to calculate total number of features on demand (only if needed). This allows formats
+     * that do not need the total to avoid calculating this expensive result, it also may be that
+     * some data stores can better estimate this total is obtained after traversing results.
+     *
+     * @param totalNumberOfFeatures Delayed calculation of total number of featuers.
+     */
+    public void setLazyTotalNumberOfFeatures(Supplier<BigInteger> totalNumberOfFeatures) {
+        if (lazyTotalNumberOfFeatures != null) {
+            throw new IllegalStateException("Total number of features has already been set");
+        }
```

Comment:
No need to check this, especially if you remove the getter for the supplier, which does not seem to be required anyways.